sycl/include/sycl/ext/oneapi/experimental/throttled_wait.hpp

```diff
@@ -0,0 +1,72 @@
+//===------- throttled_wait.hpp -  sleeping implementation of wait   ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <chrono>
+#include <thread>
+
+// The throttled_wait extension requires the inclusion of this header.
+// If we instead want it to be included with sycl.hpp, then this defnition
+// will need to be removed from here and
+// added to llvm/sycl/source/feature_test.hpp.in instead.
+#define SYCL_EXT_ONEAPI_THROTTLED_WAIT 1
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext::oneapi::experimental {
+
+template <typename Rep, typename Period>
+void ext_oneapi_throttled_wait(
+    sycl::event &e, const std::chrono::duration<Rep, Period> &sleep) {
+  while (e.get_info<sycl::info::event::command_execution_status>() !=
+         sycl::info::event_command_status::complete) {
+    std::this_thread::sleep_for(sleep);
+  }
+  e.wait();
```

Comment:
This isn't guaranteed to work.  From [the specification](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_forward_progress):
> SYCL commands submitted to a queue are not guaranteed to begin executing until a host thread blocks on their completion. In the absence of multiple host threads, there is no guarantee that host and device code will execute concurrently.

Polling on the event status could put an application into an infinite loop, because you'll never reach the call to wait.

Reply:
Would it be sufficient to call `queue::ext_oneapi_prod()` on the associated queue prior to the polling?

Reply:
Unfortunately, no. `prod()` is also defined as a hint, and doesn't provide a strong guarantee that anything will actually start executing.

Pretty much everything related to the forward progress of the device as a whole is currently defined as a hint, because there are valid implementations (e.g., SimSYCL) where everything executed by the "device" is actually executed by the host thread which eventually calls `wait`.

Being able to reason about cases where a device could execute kernels concurrently with the host thread and/or request for that to happen would require some new extension work.

Reply:
> Polling on the event status could put an application into an infinite loop, because you'll never reach the call to wait

This might not necessarily be the case.  The spec wording you quote above is true in general, but the code being added here only needs to work for the DPC++ implementation.  Does DPC++ already have a guarantee that commands will start executing even before `wait` is called?  If not, we could add an internal function call here that does provide that guarantee.

Reply:
> Does DPC++ already have a guarantee that commands will start executing even before `wait` is called? If not, we could add an internal function call here that does provide that guarantee.

Honestly, I'm not sure.  I'm worried that the answer is really complicated, though, and depends on a bunch of configuration options.

OpenCL has similar wording to SYCL regarding the guarantees about when kernels execute, so I don't think DPC++ can provide that guarantee when running on the OpenCL backend.  Our OpenCL implementation for GPUs used to batch kernels before execution, and unless that's changed recently I don't think kernels are guaranteed to begin execution.  Our OpenCL implementation for CPUs has a mode where kernel execution begins immediately on a pool of TBB threads, and the host thread simply joins the pool when it reaches `wait`, but I don't know if that's the default.

For Level Zero, it will depend on the value of `SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS`, which according to the documentation takes different default values for Windows and Linux.

For the native CPU backend, I don't know for sure.  Their behavior might be the same as TBB's above, or they might wait until `wait` to use all the logical cores for kernel execution.

For CUDA and HIP, I have no idea. I suspect that submitted kernels always begin executing on the GPU in practice, but I don't know if this is actually guaranteed by the runtime or not.